[caando] iP

caando opened the pull request:
# DukePro  

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes)) 

DukePro frees your mind of having to remember things you need to do. It's, 

- text-based
- UI from the 1990s
- easy to learn
- <del>FAST</del> _SUPER_ FAST to use

---

All you need to do is,

1. Download it from [here](https://github.com/caando/ip/releases/tag/A-Jar). 
2. Double-click it. 
3. Add your tasks. 
4. Let it manage your tasks for you 😉 

And it is **FREE!** (unless you want to donate money to me)

---

### Features  

- [X] Managing tasks
- [X] Managing deadlines
- [X] Managing events
- [ ] Reminders (coming soon)

---

### Java Practice  

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:  

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

caando committed:
Merge branch 'add-gradle-support' into branch-A-Gradle

caando committed:
feat: Setup gradle

caando committed:
Merge branch 'branch-A-Gradle'

caando committed:
feat: Add checkstyle and suppressions config

caando committed:
feat: Add checkstyle gradle task

caando committed:
docs: Add Javadoc for Duke

caando committed:
Merge branch 'branch-A-CheckStyle'

caando committed:
feat: Rename, Greet, Exit

caando committed:
feat: Echo

caando committed:
feat: Add, List

caando committed:
feat: Add task

caando committed:
feat: Add TaskList

caando committed:
feat: Add command interface and invalid command

caando committed:
feat: Add AddCommand

caando committed:
feat: Add ListCommand

caando committed:
feat: Add MarkCommand

caando committed:
feat: Add UnmarkCommand

caando committed:
feat: Use new commands in Duke

caando committed:
refactor: Move task creation to AddCommand

caando committed:
feat: Add Todo and TodoCommand

caando committed:
feat: Add Event and AddEventCommand

caando committed:
feat: Add Deadline and AddDeadlineCommand

caando committed:
feat: ToDos, Events, Deadlines

caando committed:
feat: Update text-ui-test

caando committed:
refactor: Standardize parsing methods

caando committed:
feat: Add parsing errors for commands

caando committed:
feat: Deprecate AddCommand and make Task abstract

caando committed:
test: Update text-ui-test for error handling

caando committed:
feat: Add DukeException

caando committed:
fix: Remove extra line separator

caando committed:
feat: Add delete

caando committed:
test: Add Delete to text-ui-test

caando committed:
feat: Add CommandType enum

caando committed:
Merge branch 'branch-Level-1_to_6'

caando committed:
feat: Move DukeException to exception package

caando committed:
feat: Move Duke to duke package

caando committed:
Merge branch 'branch-A-Packages'

caando committed:
feat: Add getTaskIcon and toPsvString for tasks and tasklist

caando committed:
feat: Add save to PSV file feature

caando committed:
Merge branch 'branch-Level-7'

caando committed:
feat: Add dates to deadline and event tasks

caando committed:
fix: Change text-ui-test to use dates

caando committed:
Merge branch 'branch-Level-8'

caando committed:
refactor: Move commands, tasks and exceptions to duke package

caando committed:
feat: Add TaskContainer interface

caando committed:
feat: Add more specific exceptions

caando committed:
feat: Add fromPsvString for tasks

caando committed:
feat: Add Cli

caando committed:
feat: Add FileStorage

caando committed:
feat: Add Cli with interface

caando committed:
feat: Add FileStorage with interface

caando committed:
feat: Add TaskNotFoundException

caando committed:
feat: Add date utils

caando committed:
feat: Update tasks

caando committed:
feat: Use regex, ui and storage for commands

caando committed:
feat: Add parser

caando committed:
feat: Update text-ui-test

caando committed:
Merge branch 'branch-A-MoreOOP'

caando committed:
feat: Add javadoc for commands

caando committed:
doc: Add javadoc for exceptions

caando committed:
doc: Add javadoc for storage

caando committed:
doc: Add javadoc for ui

caando committed:
doc: Add javadoc for tasks

caando committed:
doc: Add javadoc for Duke, Utils and Parser

caando committed:
fix: Update expected data for text-ui-test

caando committed:
style: Remove trailing whitespace

caando committed:
Merge branch 'branch-A-JavaDoc'

caando committed:
feat: Add start to Ui

caando committed:
test: Add CliTest

caando committed:
fix: Update Duke to start ui

caando committed:
test: Add ParserTest

caando committed:
feat: Add getters for AddDeadlineCommand

caando committed:
test: Add tests for AddDeadlineCommand

caando committed:
style: Fix style for tests

caando committed:
test: Add AddEventCommand test

caando committed:
style: Fix style

caando committed:
test: Add AddTodoCommandTest

caando committed:
test: Add DeleteCommandTest

caando committed:
test: Update text-ui-test

caando committed:
Merge branch 'branch-A-JUnit'

caando committed:
feat: Add FindCommand

caando committed:
style: Remove trailing spaces

caando committed:
Merge branch 'branch-Level-9'

tadacheng commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,85 @@
+     * @throws ParseCommandException If the input is invalid or cannot be parsed.
+     */
+    public static Command parse(String input) throws ParseCommandException {
+        String regex = "find\\s+(.+)";
```
I like your use of regex to parse the user inputs, if the regex pattern is the same maybe you can make it static?

caando replied:
Good point!

tadacheng commented on src/main/java/duke/command/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,106 @@
+     *
+     * @return the task description
+     */
+    public String getTaskDescription() {
```
I saw that this is only used in testing, should this be moved to a test utility class?

tadacheng commented on src/main/java/duke/ui/Cli.java:
```diff
@@ -0,0 +1,132 @@
+     * Displays a goodbye message and a separator line before closing the interface.
+     */
+    @Override
+    public void close() {
```
Perhaps you can add scanner.close() to prevent potential resource leaks?

caando replied:
You're right, thanks for reminding!

tadacheng commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,60 @@
+ * and user interface, then enters an interactive loop where user commands are parsed, executed, and handled.
+ * The application continues running until the user inputs the "bye" command, at which point the program terminates.
+ */
+public class Duke {
```
I like your JavaDocs, it clearly explains the purpose of each method. 
I also did not find any issue with your coding standard

aisyahally commented on src/main/java/duke/exception/InvalidStatusIconException.java:
```diff
@@ -0,0 +1,33 @@
+package duke.exception;
```
I like that there are many different classes of Exceptions. It makes it clearer what errors you are catching when I read the other parts of the code.

aisyahally reviewed (commented):
Overall, I could not find any errors in coding standards or code naming. Good job! It was readable and clear. The allocation of functions to the different classes helped in following the code and I liked how you used regex to match the patterns.

caando committed:
feat: Add fxml files for GUI

caando committed:
feat: Add profile images

caando committed:
feat: Add DialogBox

caando committed:
feat: Add MainWindow

caando committed:
feat: Add Duke class with process command method

caando committed:
feat: Add GuiAdaptor for Duke

caando committed:
feat: Add Gui application and launcher

caando committed:
feat: Add gradle package and made Gui nicer

caando committed:
Merge branch 'branch-Level-10'

caando committed:
fix: Make add commands save to storage

caando committed:
fix: Add mockito to gradle dependancy

caando committed:
style: Remove start of line parenthesis

caando committed:
fix: Close scanner to prevent resource leaks

caando committed:
fix: Use correct launcher class for gradle

caando committed:
style: Make regex patterns static and final

caando committed:
Merge branch 'branch-Bugfixes-and-Styling'

caando committed:
feat: Add assertions for command params

caando committed:
feat: Add input string assertions

caando committed:
feat: Add state assertions for storage and ui

caando committed:
refactor: Improve task iteration for clarity and standardization

The current task iteration in `FileStorage` uses a lambda approach,
adding complexity. `TaskContainer` also lacks native iteration support.

Replaced the lambda with a for-each loop and made `TaskContainer`
implement `Iterable<Task>`, adding an `iterator()` to `TaskList`.

This simplifies the code, aligns with Java best practices, and improves
maintainability without changing functionality.

caando committed:
refactor: Use streams for task filtering and list generation

Previously, task filtering and list output generation relied on manual
iteration, making the code verbose and harder to maintain.

Replaced manual loops with Java Streams and `IntStream`, improving code
readability and efficiency. Task indices are now mapped directly to list
elements, and error handling is integrated into the stream logic.

This approach simplifies logic, aligns with modern Java practices, and
maintains correct task indexing without functionality changes.

caando committed:
feat: Add gradle.yml template

caando committed:
fix: Update runtest.bat to match runtest.sh

caando committed:
fix: Fix compile command in runtest.bat

caando committed:
fix: Replace wildcards with recursive loop

caando committed:
fix: Use correct class package

caando committed:
fix: Use system line separator for Cli test

caando committed:
feat: Add copy to task container

caando committed:
feat: Add duke program state

caando committed:
feat: Change command execution to use `State` object

All commands have been updated to take a unified `State` parameter instead of individual task, storage, and UI components.

### Current Situation
Previously, commands received multiple arguments (`TaskContainer`, `Storage`, and `Ui`) to execute operations. This approach made it harder to maintain and extend the codebase.

### Why the Change Was Needed
- Reduces method signature complexity and improves readability
- Simplifies future extensions by centralizing shared data access through a single `State` object
- Aligns code with cleaner object-oriented design principles

### Changes Made
- Introduced a new `State` object in the execution flow
- Updated all command implementations, including `AddDeadlineCommand`, to unpack components from `State`
- Ensured backward compatibility by retaining task state updates

This refactor makes the code more scalable and maintainable by adhering to encapsulation best practices and reducing redundancy.

caando committed:
feat: Update duke process to use state

caando committed:
test: Update tests based on new command interface

caando committed:
feat: Add previous input string for mutable commands

caando committed:
feat: add UndoCommand to revert the last state change

Add `UndoCommand` to enable users to revert to the previous application state if available.

- Implements `UndoCommand#execute` to restore the previous state.
- Displays a message when no prior state exists or when an undo is successful.
- Follows the existing `Command` structure and design patterns.

This feature improves user experience by allowing error recovery and greater flexibility when managing tasks.

caando committed:
Merge pull request #2 from caando/branch-A-CI

feat: A-CI

caando committed:
Merge pull request #4 from caando/branch-A-Assertions

refactor: A-Assertions

caando committed:
Merge branch 'master' into branch-A-Streams

caando committed:
Merge pull request #1 from caando/branch-A-Streams

refactor: A-Streams

caando committed:
Merge branch 'master' into branch-C-Undo

caando committed:
Merge pull request #3 from caando/branch-C-Undo

feat: C-Undo

caando committed:
docs: Fix javadoc errors, warnings and style

caando committed:
refactor: Make happy path prominent

caando committed:
refactor: Remove unused

caando committed:
Revert "refactor: Remove unused"

This reverts commit 3e779e3f868dd3b4f1c340382e3e136d4d24bd6c.

caando committed:
fix: Correct Typo

caando committed:
docs: Fix formatting

caando committed:
Merge branch 'branch-A-CodeQuality'

caando committed:
feat: Add angry Meeseeks for error

caando committed:
feat: Add text box coloring

caando committed:
feat: Add dynamic screen sizing

caando committed:
Merge branch 'branch-A-BetterGui'

caando committed:
feat: Add Meeseeks personality to commands

caando committed:
test: Update tests for new personality

caando committed:
Merge branch 'branch-A-Personality'

caando committed:
feat: Add user guide

caando committed:
feat: Add demo image to user guide

caando committed:
feat: Add title to Gui

caando committed:
Merge branch 'branch-A-BetterGui'

caando committed:
Merge branch 'master' into branch-A-UserGuide

caando committed:
docs: Update demo image with title

caando committed:
docs: Update title in readme.md

caando committed:
Merge branch 'branch-A-UserGuide'

caando committed:
feat: Add AI.md

caando committed:
Merge branch 'branch-A-AiAssisted'

caando committed:
docs: Rename Demo.png to Ui.png

caando committed:
fix: Fix undo for mark and unmark

caando committed:
Merge branch 'branch-Bugfixes-and-Styling'

caando committed:
fix: Change error to output on missing file

caando committed:
Merge branch 'branch-Bugfixes-and-Styling'

caando committed:
docs: Improve userdoc styling

caando committed:
feat: Add bye command and storage details to userdoc

caando committed:
Merge branch 'branch-A-UserGuide'
